### PR TITLE
Extend sort's floating point optimizations to Float16

### PR DESCRIPTION
**Pull request by LilithHafner**
Previously, we treated Float16 as an unknown type which we use the standard `lt` function to deal with. This PR extends the optimizations applied to `Float32` and `Float64` to also apply to `Float16`. These optimizations are especially nice because with so few bits, `Float16`s are easily sorted by counting sort or radix sort. The net result is a 30x speedup for large vectors and a smaller speedup for smaller vectors. The increase in allocations is due to use of a different algorithm. The fpsort optimizations are phased out [for vectors smaller than 10](https://github.com/JuliaLang/julia/blob/fb760d973630afc7560be4c225a4a2b2ad1b9fa5/base/sort.jl#L1578-L1580) when they stop being helpful.

<details>

<summary> Benchmarks: </summary>

```julia
for i in 1:8
    x = rand(Float16, 10^i)
    @time sort!(x)
end

# before
  0.000001 seconds
  0.000005 seconds
  0.000081 seconds
  0.000883 seconds
  0.008984 seconds
  0.120891 seconds
  0.684031 seconds
  8.030174 seconds

# after
  0.000003 seconds
  0.000006 seconds (3 allocations: 624 bytes)
  0.000013 seconds (3 allocations: 3.156 KiB)
  0.000101 seconds (4 allocations: 20.703 KiB)
  0.000466 seconds (2 allocations: 120.047 KiB)
  0.004767 seconds (2 allocations: 120.047 KiB)
  0.036061 seconds (2 allocations: 120.047 KiB)
  0.297907 seconds (2 allocations: 120.047 KiB)

for i in 1:8
    @btime sort!(x) setup=(x=rand(Float16, 10^$i)) evals=1
end

# before
  171.000 ns (0 allocations: 0 bytes)
  2.724 μs (0 allocations: 0 bytes)
  40.461 μs (0 allocations: 0 bytes)
  538.888 μs (0 allocations: 0 bytes)
  5.834 ms (0 allocations: 0 bytes)
  63.062 ms (0 allocations: 0 bytes)
  878.109 ms (0 allocations: 0 bytes)
  12.620 s (0 allocations: 0 bytes)

# after
  189.000 ns (0 allocations: 0 bytes)
  1.476 μs (3 allocations: 624 bytes)
  6.955 μs (3 allocations: 3.16 KiB)
  66.651 μs (4 allocations: 20.70 KiB)
  333.517 μs (2 allocations: 120.05 KiB)
  2.643 ms (2 allocations: 120.05 KiB)
  29.109 ms (2 allocations: 120.05 KiB)
  293.027 ms (2 allocations: 120.05 KiB)
```

</details>

### CODE REVIEW

**Comment by LilithHafner** [comment]
@oscardssmith Would you mind reviewing this, or would you prefer I bug someone else about my sorting PRs?

File: base/sort.jl [hunk 1]
```diff
@@ -1464,10 +1464,12 @@ import Core.Intrinsics: slt_int
...
 import ...Order: lt, DirectOrdering
 
-const Floats = Union{Float32,Float64}
-const FPSortable = Union{ # Mixed Float32 and Float64 are not allowed.
+const Floats = Union{Float16, Float32, Float64} # IEEEFloat is not available in Core.Compiler
+const FPSortable = Union{ # Mixed bitwidths are not allowed.
+    AbstractVector{Union{Float16, Missing}},
     AbstractVector{Union{Float32, Missing}},
     AbstractVector{Union{Float64, Missing}},
+    AbstractVector{Float16},
```
Comment (oscardssmith) [inline base/sort.jl:1472]: maybe `AbstractVector{<:IEEFloat}`

**Comment by oscardssmith** [comment]
Overall, looks good!

File: base/sort.jl (hunk 1 above)
Comment (LilithHafner) [reply base/sort.jl:1472]: That would include Vector{Union{Float32, Float64}} which is not safe to compare via `slt_int` (see https://github.com/JuliaLang/julia/issues/42739). I'll link to that issue in a comment.

Comment (oscardssmith) [reply base/sort.jl:1475]: makes sense.

**Review by oscardssmith — APPROVED** [review]

**Pull request merged**